### PR TITLE
Flake8

### PR DESCRIPTION
**Pull request by ocefpaf**
Looks a lot but most of it is `pep8`,  `typos`, and Python 2 and 3 compliant code.

The only code changed is from `pyselfe_kdtree.py`, but if I had to guess that code is currently broken and my changes did not fix it.

What needs some review:

- changed how the version number is obtained.
- some re-wording in the `README.rst`
- removed Python 3.3 test (They still pass, but we should discourage py33 use.)
- enforce `flaky8` tests on Travis-CI

The last one is to avoid #114.

### CODE REVIEW

**Comment by ChrisBarker-NOAA** [comment]
where are we on this? as you say --mostly pep8 stuff.

Though I don't think it's a good idea to enforce flake8 -- I really don't like being pedantic about that: " a foolish consistency is the hobgoblin of little minds" and all that.

Let's just agree to be adults here. :-)

**Comment by ChrisBarker-NOAA** [comment]
Oh, I see this in setup.py:

        'Programming Language :: Python :: 3',
        'Programming Language :: Python :: 3.3',
        'Programming Language :: Python :: 3.4',

should that be 3.4 and 3.5, and not 3.3?

**Comment by ocefpaf** [comment]
> should that be 3.4 and 3.5, and not 3.3?

Fixed.

> where are we on this? as you say --mostly pep8 stuff.

Mostly avoid #114 cases and to make the code friendlier to future contributions.

> Though I don't think it's a good idea to enforce flake8 -- I really don't like being pedantic about that: " a foolish consistency is the hobgoblin of little minds" and all that.

We can skip some rules and/or files.  Also, we can always add `noqa` for long line.
I strongly believe that we need to enforce.  The advantages (cleaner code, easy reviews, catch bad syntax, lower entry barrier to read the code, etc) outweigh the disadvantages (minor code corrections).

> Let's just agree to be adults here. :-)

Don't think that is the case.  The more the contributor base grows the more we need to enforce a few rules.  Pyugrid will became a target for a larger community soon.

**Comment by ChrisBarker-NOAA** [comment]
OK, if you insist :-)

Ideally, linters would be run when you commit, not when you build, but whatever.

But see the conversation on python-dev about this -- almost no one thinks enforcing a linter is a good idea.

I run Flake8 inside sublime when I'm teaching (I guess I need to turn that on on my dev machine), but I've ended up turning off quite a few of the checks -- they just annoy me very unnecessarily.

If nothing else we should extend the line-length limit -- I think an 83 character line (for instance) is far more readable than one chopped up -- maybe 90 characters or so?

**Comment by ocefpaf** [comment]
> If nothing else we should extend the line-length limit -- I think an 83 character line (for instance) is far more readable than one chopped up -- maybe 90 characters or so?

I am using 100.  (See https://github.com/pyugrid/pyugrid/pull/116/files#diff-354f30a63fb0907d4ad57269548329e3R25.)

> Ideally, linters would be run when you commit, not when you build, but whatever.

Agreed. The testing is only to catch people who don't :wink:

Also, the testing on Travis-CI is not testing on build time.  People can still build without `flake8`.  That test help us review the PRs.  Nothing more.

Note that, If a code is non-compliant and you still want to merge, we need to add an ignore statement on Tavis-CI and merge it regardless.

**Comment by ChrisBarker-NOAA** [comment]
OK -- then sure, turn it on.  
I suspect we'll have to go in and turn off some of the more pedantic checks once we start using it, but you're right, it will make for cleaner code.

-CHB

**Comment by ChrisBarker-NOAA** [comment]
slight conflict with changes I had just made -- but all good now.

**Comment by PythonCHB** [comment]
OK, now that we've got the flake8 checker, my commits failed :-(

My bad, but I am wondering if I really care. Clearly, I have standard practices that violate pep8 ;-)

I'll try to fix those soon. But I'm not sure I like that it comes back as a failure from Travis. It took a bit of poking to see that it was just flake8 that failed.

**Comment by ChrisBarker-NOAA** [comment]
OK,

I am really not liking this.

Sure, a bunch of it is a bit cleaner than it was, but:

flake8 wants: arr[:, 1]
frankly, I prefer: arr[:,1] without the space after the comma.

but I do like spaces after the commas in other contexts.

And the killer: I getting:
./test/test_io_util.py:7:1: F401 'load_from_varnames' imported but unused

which is true -- it isn't used. But in this case, I haven't written any proper tests for that code, but I do want to make sure that it at least imports.

This is all why I want linters to chastise me, but not actually indicate failure.

**Comment by ocefpaf** [comment]
> which is true -- it isn't used. But in this case, I haven't written any proper tests for that code, but I do want to make sure that it at least imports.

Good practices would say that you should only commit when you have those tests :wink: 

@ChrisBarker-NOAA, to be honest, I don't like it either!  However, `pyugrid` needs to stop being a niche grid parser developed by 2-3 people and become a tool for a large community.

I strongly believe that we will only achieve this thorough good practices and clean code.

**Comment by rsignell-usgs** [comment]
This is tough love, but I think @ocefpaf is right here.    Karl Taylor is interested in using UGRID for CMIP6, so the more professional we can be, the better.

**Comment by ChrisBarker-NOAA** [comment]
I have have no problem with coding standards and clean code -- great stuff!

But I don't think that it's good to be a slave to automated linting tools.
They can really get in the way.

Again, take a look at python-dev recently -- someone proposed automatically
running a linter on.

Failing tests are a much bigger deal than a missing space after a comma. In
fact, in the case of a comma used for indexing multi-dimensional arrays --
a case pep8 hadn't even considered, I think the pep8 compliant way makes it
harder to read. And there are other more special cases where you don't want
to be pep8 compliant, too. The core problem is that while style is
important, some aspects of style are far more important than another.

If there is a way to gently tell contributors that they need to clean up
their code without showing it as failed tests, then I'm all for it.

-CHB



On Sat, Feb 13, 2016 at 3:23 AM, Rich Signell <notifications@github.com>
wrote:

> This is tough love, but I think @ocefpaf <https://github.com/ocefpaf> is
> right here. Karl Taylor is interested in using UGRID for CMIP6, so the more
> professional we can be, the better.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyugrid/pyugrid/pull/116#issuecomment-183645895>.
>



-- 

Christopher Barker, Ph.D.
Oceanographer

Emergency Response Division
NOAA/NOS/OR&R            (206) 526-6959   voice
7600 Sand Point Way NE   (206) 526-6329   fax
Seattle, WA  98115       (206) 526-6317   main reception

Chris.Barker@noaa.gov

**Comment by ocefpaf** [comment]
> But I don't think that it's good to be a slave to automated linting tools.

@ChrisBarker-NOAA you mention Python-dev as an example that this practice is bad.  I agree, but I don't believe it applies to our case here.  Earth scientists are not coders who will infringe recommended practices with a good reason.  You, Kyle, and many good coders I see contributing to libraries do **know** what you are doing, but people like me, @rsignell-usgs, and many more future contributors **will** benefit from this test.

Besides, the code base had many violations that scrambled the code and made it hard to follow.  Not to mention syntax erros that were passed silently!  Adding the `flake8` test will help the review process and prevent that from happening again.  Virtually every project I know and contribute to uses such a test. (Iris, Cartopy, PyCSW, OWSLib, all of Axiom/ASA modern code, etc).

> They can really get in the way.

I still think that the advantages outweighs the disadvantages.  Besides, it is only 1 test that servers as a reviewer helper-tool. We **can**  merge good code that fails this test and passes the others.  We only need to check the Travis-CI log and use our best judgment.

If you want to remove that go ahead.  I am not against it.  I  think it will make reviewing process harder and/or flawed. (Something like #114 would never happen with this test on.)

**Comment by rhattersley** [comment]
> It took a bit of poking to see that it was just flake8 that failed.

See #124 for a way to make this easier.
